### ISSUE TITLE
How to access dynamic DbSets that were created by the builder

### ISSUE THREAD

**Issue by idan-h** [dotnet/efcore#26636]
Using the latest version of EF Core with C#  
  
I am creating DbSets dynamically and that works fine:  

```
public class ApplicationDbContext : DbContext
    {
        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options) : base(options)
        {
        }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);

            foreach (var model in ModelHelper.AllModels)
            {
                // Create the table
                builder.Entity(model).ToTable(ModelHelper.GetTableName(model));

                // Sets the identity column
                builder.Entity(model).Property<int>(nameof(BaseModel.Id)).UseIdentityColumn();

                // Sets the auto datetime of the DateInserted column
                builder.Entity(model).Property<DateTime>(nameof(BaseModel.DateInserted)).HasDefaultValueSql("NOW()");
            }
         ......

        }
    }
```  
 
Now to access the model, I'm using Dapper and building the queries myself because I haven't found any other viable way to access the models.  

I tried using `dbContext.Set<T>()` but the following error is being raised:  
```
InvalidOperationException: Cannot create a DbSet for 'Model' because this type is not included in the model for the context.  
 ```
I'm guessing that in order to use the DbSet I must staticlly type the `DbSet<>` inside `ApplicationDbContext`.  
  

My question is if there's another way to access the dbSet instead of dynamically creating the queries myself.  
  
Thank You!

**Comment by roji** [comment]
There's no need to have DbSet properties on your context type - just using `DbSet<T>` works just fine. Can you please post a runnable code sample that shows exactly what you're doing?

**Comment by idan-h** [comment]
> There's no need to have DbSet properties on your context type - just using `DbSet<T>` works just fine. Can you please post a runnable code sample that shows exactly what you're doing?

I have this type:
```
public class ConnectionLog : BaseModel
    {
        public Client Client { get; set; }
        public string LocalUser { get; set; }
        public string IpAddress { get; set; }
    }
```  

that is being added to the builder in OnModelCreating as written in the code above.  
Then, if I try to call this:
```
public class IndexModel : PageModel
    {
        private readonly ApplicationDbContext _context;

        public IndexModel(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<IActionResult> OnGet()
        {
            var set = _context.Set<Client>();
            await set.AsQueryable().ToListAsync();

            return Page();
        }

    }
```  

This error is being raised:
```
InvalidOperationException: Cannot create a DbSet for 'Model' because this type is not included in the model for the context.
```

**Comment by roji** [comment]
@idan-h you have included the ConnectionLog entity type in your model, so it's not aware of it. It's true that this can be done via statically-typed DbSet properties on the model, but you can also do this by adding `modelBuilder.Entity<ConnectionLog>();` to your model's OnModelCreating.

This is covered in [these docs](https://docs.microsoft.com/en-us/ef/core/modeling/entity-types?tabs=data-annotations#including-types-in-the-model) - I suggest giving that a read.

We don't have a non-generic `Set(Type)` at this time, because that would be very hard to use for actual queries (LINQ is strongly-typed). However, if you absolutely want to do this, you can always use reflection via `typeof(DbSet).MakeGeneric(typeof(ConnectionLog))`. But you still have to include ConnectionLog in your model as above.

**Comment by idan-h** [comment]
> @idan-h you have included the ConnectionLog entity type in your model, so it's not aware of it. It's true that this can be done via statically-typed DbSet properties on the model, but you can also do this by adding `modelBuilder.Entity<ConnectionLog>();` to your model's OnModelCreating.
> 
> This is covered in [these docs](https://docs.microsoft.com/en-us/ef/core/modeling/entity-types?tabs=data-annotations#including-types-in-the-model) - I suggest giving that a read.
> 
> We don't have a non-generic `Set(Type)` at this time, because that would be very hard to use for actual queries (LINQ is strongly-typed). However, if you absolutely want to do this, you can always use reflection via `typeof(DbSet).MakeGeneric(typeof(ConnectionLog))`. But you still have to include ConnectionLog in your model as above.

I am using the non generic `builder.Entity(type)`, does that make a difference? If so, what is the difference between them?  

  
And what abouy using linq queries with a dynamic type, so it would be interpeted at runtime?

**Comment by roji** [comment]
> I am using the non generic builder.Entity(type), does that make a difference? If so, what is the difference between them?

The difference is that builder.Entity is used to configure your model, whereas DbSet is used to execute queries. The latter is done by composing queryable LINQ operators, which are themselves generic; but model configuration doesn't necessarily require any generics. Try thinking what you would actually do with a non-generic DbSet - it's simply not a very useful thing.

**Comment by idan-h** [comment]
> > I am using the non generic builder.Entity(type), does that make a difference? If so, what is the difference between them?
> 
> The difference is that builder.Entity is used to configure your model, whereas DbSet is used to execute queries. The latter is done by composing queryable LINQ operators, which are themselves generic; but model configuration doesn't necessarily require any generics. Try thinking what you would actually do with a non-generic DbSet - it's simply not a very useful thing.

Well, you could do what I wrote above (when you certainly know that you have specific property, and want to filter by that).
Also, you can write strings that would be parsed as lambda expression. That lets you the option of making things dynamic instead of statically typed and it can be useful for example when you want to build a simple generic CRUD operations platform without writing many views. This is what I am using it for, you can just add a new model, derive from a base class, and you immediately manage your model without any other configuration needed. (a single view, no need to write specific code for each model).  
  
If you don't find it important enough, it's ok, but it can be a great time saver for a lot of generic and simple needs.

Thank you anyway!

**Comment by roji** [comment]
> Also, you can write strings that would be parsed as lambda expression.

How do you do that? I'm assuming you mean that your application will parse some user-provided string query, and dynamically build query operators via reflection? If you do that, that means you're going to need to use reflection to generate generic LINQ queryable operators like [Where](https://docs.microsoft.com/en-us/dotnet/api/system.linq.queryable.where?view=net-5.0#System_Linq_Queryable_Where__1_System_Linq_IQueryable___0__System_Linq_Expressions_Expression_System_Func___0_System_Boolean___) - because there's no non-generic where. That's certainly possible, but at that point you may as well just generate the generate DbSet as well (again, via `typeof(DbSet).MakeGeneric(typeof(ConnectionLog))`).

To summarize, at the end of the day querying uses generic query operators, so it doesn't seem useful to have a non-generic DbSet.

**Comment by idan-h** [comment]
> > Also, you can write strings that would be parsed as lambda expression.
> 
> How do you do that? I'm assuming you mean that your application will parse some user-provided string query, and dynamically build query operators via reflection? If you do that, that means you're going to need to use reflection to generate generic LINQ queryable operators like [Where](https://docs.microsoft.com/en-us/dotnet/api/system.linq.queryable.where?view=net-5.0#System_Linq_Queryable_Where__1_System_Linq_IQueryable___0__System_Linq_Expressions_Expression_System_Func___0_System_Boolean___) - because there's no non-generic where. That's certainly possible, but at that point you may as well just generate the generate DbSet as well (again, via `typeof(DbSet).MakeGeneric(typeof(ConnectionLog))`).
> 
> To summarize, at the end of the day querying uses generic query operators, so it doesn't seem useful to have a non-generic DbSet.

I see what you mean.  
What I meant was that you can use libraries like https://dynamic-linq.net/  
to create the linq via a string dynamically, and then you have more dynamic functionality (creating queries without the need of knowing the type before compiling) because you don't need to use a generic type.   
Or maybe parse the linq dynamically yourself (compile the expression)

But yea, this can be done with reflection right now anyway.  

Thank you again
